### PR TITLE
+ fix-up issue #143 (type mismatch in LLVM CodeGen)

### PR DESCRIPTION
**Pull request by wonsubkim**
- In LLVM CodeGen, missed interop type checking
 Fixed it as if Sema does
 - In function using interop type, interop type information is stored within function prototype
 while checking if actual argument type(RHS) matches function formal parameter type(LHS),
 it also consider that unchecked pointer type can have interop type
  - check if function parameter has related interop type
  - check if interop type as well as original type matches argument type
 - make check-clang, make check-checked passed
  - failing testcases are passed after fix-up

### CODE REVIEW

File: lib/CodeGen/CodeGenFunction.h [hunk 1]
```diff
@@ -3317,6 +3317,13 @@ class CodeGenFunction : public CodeGenTypeCache {
 
   template<typename T>
   static bool isObjCMethodWithTypeParams(const T *) { return false; }
+
+  // Determine whether the given argument is FunctionProtoType
+  // that have parameter bounds information within it
+  static bool isFunctionProtoTypeWithTypeParams(const FunctionProtoType* FPT) { return true; }
```
Comment (dtarditi) [inline lib/CodeGen/CodeGenFunction.h]: I think this should be renamed to isFunctionProtoType - there's really nothing about type parameters here.

File: lib/CodeGen/CodeGenFunction.h [hunk 2]
```diff
@@ -3333,22 +3340,43 @@ class CodeGenFunction : public CodeGenTypeCache {
     if (CallArgTypeInfo) {
 #ifndef NDEBUG
       bool isGenericMethod = isObjCMethodWithTypeParams(CallArgTypeInfo);
+      bool isFunctionProtoType = isFunctionProtoTypeWithTypeParams(CallArgTypeInfo);
```
Comment (dtarditi) [inline lib/CodeGen/CodeGenFunction.h]: You'll need to rename this variable to not conflict with the updated name.

File: lib/CodeGen/CodeGenFunction.h [hunk 3]
```diff
@@ -3333,22 +3340,43 @@ class CodeGenFunction : public CodeGenTypeCache {
...
 #endif
 
       // First, use the argument types that the type info knows about
+      int N = 0;
       for (auto I = CallArgTypeInfo->param_type_begin() + ParamsToSkip,
                 E = CallArgTypeInfo->param_type_end();
-           I != E; ++I, ++Arg) {
+           I != E; ++I, ++Arg, ++N) {
         assert(Arg != ArgRange.end() && "Running over edge of argument list!");
+        const Type* ty1 = getContext().getCanonicalType((*I).getNonReferenceType()).getTypePtr();
```
Comment (dtarditi) [inline lib/CodeGen/CodeGenFunction.h]: All this code should be under ifndef NDEBUG.   We only want this code to run if the assert will be executed.

**Comment by lenary** [comment]
Hi @wonsubkim, you mention you have passing testcases. Please can you submit them as a pull request so I can run your branch against them too, and make sure they haven't missed anything?

**Comment by lenary** [comment]
Also, follow-up, @wonsubkim please may you make sure that you make a branch for each pull-request, rather than committing to your repository's master branch. This will mean merging becomes significantly easier.

**Comment by wonsubkim** [comment]
since this problem is fixed by better solution, I will close this pull request
I will do PR for this testcase at my checkedc branch

**Comment by wonsubkim** [comment]
I have applied David's feedback and tested
I have completed test for release/debug mode 'make check-clang', 'make check-checkedc'

`make check-checked`
release mode : unexpected error (0)
debug mode : unexpected error (0)
In make check-checkedc, I have added simple LLVM code generation error case into checkedc test cases
It has passed successfully

`make check-clang`
release mode : unexpected failure (1) : CheckedC/dynamic-checks/array-subscript-code-gen.c (1103 of 9830)
debug mode : unexpected failure(1) : CheckedC/dynamic-checks/array-subscript-code-gen.c (1103 of 9803)

In make check-clang test, there is only one failure case in both mode
However, I think that error case is not relevant to interoperation type

I will check out issue-143 branch into my private fork repository

**Comment by wonsubkim** [comment]
I have followed your git workflow
I have applied all changes into existing-work branch
I have reset master and sync master with MS repository
I have created a new branch for 143 issue (type mismatch in LLVM code generation)
I have cherry-picked version in which feedback is applied in existing-work branch and then merged those changes into issue-143 branch
Finally, I have pushed issue-143 branch into my private fork of github
In this pull request, I can't see changes

Is it ok or what should I do?

**Comment by dtarditi** [comment]
Yes, you should create a new pull request for this issue.    You have changed branches for the work.

**Comment by wonsubkim** [comment]
Okay, I will close this PR and open another PR with different branches

**Pull request closed without merging**